shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs

```diff
@@ -120,18 +124,49 @@ genTx ::
   LedgerEnv ->
   (UTxOState era, DPState era) ->
   Gen (Tx era)
-genTx ge@(GenEnv _ (Constants {genTxRetries})) =
-  genTxRetry genTxRetries ge
+genTx ge le st =
+  tryGenTx ge le st >>= \case
+    Left isbi -> Exn.throw $ GenTxExceptionInsufficientSpendingBalance isbi
+    Right tx -> pure tx
+
+data GenTxException
+  = -- | Every transaction that the generator attempted had fees that exceeded
+    -- the available balance.
+    GenTxExceptionInsufficientSpendingBalance !InsufficientSpendingBalanceInfo
+  deriving (Show)
+
+instance Exn.Exception GenTxException
+
+-- | The data for 'GenTxExceptionInsufficientSpendingBalance'.
+data InsufficientSpendingBalanceInfo = InsufficientSpendingBalanceInfo
+  { -- | The smallest amount by which the fees exceeded the unspent balance for
+    -- any of the transactions attempted.
+    isbiDiff :: !Coin
+  }
+  deriving (Show)
+
+-- | 'Left' means the generator could not afford the fees for any of the
+-- 'genTxRetries'-many transactions it attempted.
+tryGenTx ::
+  (HasCallStack, Era era, Mock (Crypto era)) =>
+  GenEnv era ->
+  LedgerEnv ->
+  (UTxOState era, DPState era) ->
+  Gen (Either InsufficientSpendingBalanceInfo (Tx era))
```

Comment:
I feel if we have the `GenTxException` type then we should throw that. Seems weird to wrap it at the level of `genTx` whereas `tryGenTx` returns only one specific type.

Reply:
Done. Thanks!